File: IvanGame/Level.cpp
#include "Level.h"
#include "Game.h"
#include "fns.h"
#include <GL/gl.h>

Level::Level(SDL_Surface* screen) : mScreen(screen)
{
	mLoadBehavior = DoNothing;
	mUpdateBehavior = DoNothing;
}

Level::Level(SDL_Surface* screen, string n, Behavior loadBehave, Behavior updateBehave) : mScreen(screen), mName(n), mBackground(NULL)
{
	/// Set behavior
	mLoadBehavior = loadBehave;
	mUpdateBehavior = updateBehave;
}

Level::~Level()
{
	for (size_t i = 0; i < mSprites.size(); ++i) {
		delete mSprites[i];
	}
}

void Level::DrawIMG(SDL_Surface *img, int x, int y)
{
    SDL_Rect dest;
    dest.x = x;
    dest.y = y;
    SDL_BlitSurface(img, NULL, mScreen, &dest);
}


void Level::DrawIMG(SDL_Surface *img, int x, int y, int w, int h, int x2, int y2)
{
    SDL_Rect dest;
    dest.x = x;
    dest.y = y;
    SDL_Rect src;
    src.x = x2;
    src.y = y2;
    src.w = w;
    src.h = h;
    SDL_BlitSurface(img, &src, mScreen, &dest);
}


void Level::drawScene()
{
	Uint32 color;

    /** Create a black mBackgroundground using the mScreen pixel format (32 bpp) */
	glClear(GL_COLOR_BUFFER_BIT);
	
	/** Draw BG */
	if(mBackground!=NULL)
		mBackground->draw();
	
	/** Draw Sprites*/
	DrawSprites();
    
	/** Draw Boundary Data */
	DrawCollisions();

	/** Flip the working image buffer with the mScreen buffer */
	SDL_GL_SwapBuffers();
}

void Level::DrawSprites()
{
	for (size_t i=0; i<mSprites.size(); ++i) {
		mSprites[i]->draw();
	}
}

void Level::DrawCollisions()
{
	if(Game::game()->ShowCollisions){
		mBackground->drawCollisions();
		for (size_t i=0; i<mSprites.size(); ++i) {
			mSprites[i]->drawCollisions();
		}
	}
}

void Level::LoadBG(std::string name)
{
	mBackground->load(name);
}

void Level::postEvent(SDL_Event event)
{
	switch (event.type) {
	case SDL_KEYDOWN: {
		
	}
	}
}

/** Will handle all movement and dynamic triggers. */
void Level::update(){
	mUpdateBehavior();
}

void Level::load(){
	mLoadBehavior();
}


void Level::addSprite(Sprite* sp){
	mSprites.push_back(sp);
}

Sprite* Level::getSprite(string name){
	for(int i=0, count=mSprites.size(); i < count; i++){
		if(mSprites[i]->getName()==name){
			return mSprites[i];
		}
	}
	return NULL;
}

void Level::removeSprite(Sprite* sp){
	for(vector<Sprite*>::iterator ptr = mSprites.begin(); ptr != mSprites.end(); ptr++){
		if(*ptr == sp){
			mSprites.erase(ptr);
			return;
		}
	}
}

void Level::addAnimation(Animation anim){
	mAnims.push_back(anim);
}

Animation* Level::getAnimation(string name){
	for(int i=0, count=mAnims.size(); i < count; i++){
		if(mAnims[i].mName==name){
			return &mAnims[i];
		}
	}
	return NULL;
}

void Level::removeAnimation(Animation anim){
	for(vector<Animation>::iterator ptr = mAnims.begin(); ptr != mAnims.end(); ptr++){
		if(*ptr == anim){
			mAnims.erase(ptr);
			return;
		}
	}
}

void Level::addActor(string name, Actor actor){
	mActors.insert(make_pair(name, actor));
}

void Level::removeActor(string name){
	mActors.erase(name);
}

Sprite* Level::findSpriteByName(string name){
	/// \todo make this return a list of all sprites with the same name (or make it specifiable)
	for(unsigned int i=0; i < mSprites.size(); i++){
		if(mSprites[i]->getName()==name)//find the sprite with the same name
			return mSprites[i];
	}
	return NULL;//if a sprite wasn't found return null
}

File: BodilyInfection/BodilyInfection/Level.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BodilyInfection
{
    class Level
    {
        #region Constructor
        public Level()
        {
            LoadBehavior = () => { };
            UpdateBehavior = () => { };
            EndBehavior = () => { };
        }
        public Level(string n, Behavior loadBehave, Behavior updateBehave, Behavior endBehavior)
        {
            mName = n;
            LoadBehavior = loadBehave;
            UpdateBehavior = updateBehave;
        }
        #endregion Constructor

        #region Behaviors
        /// <summary>
        /// Level's load action
        /// </summary>
        public Behavior LoadBehavior { get; set; }
        /// <summary>
        /// Level's update Behavior
        /// </summary>
        public Behavior UpdateBehavior { get; set; }
        /// <summary>
        /// Level's End Behavior
        /// </summary>
        public Behavior EndBehavior { get; set; }
        #endregion Behaviors

        #region Properties
        /// <summary>
        /// Get's level's name
        /// </summary>
        public string Name { get { return mName; } }
        /// <summary>
        /// Gets and Sets Level's current Background
        /// </summary>
        public Background Background { get; set; }
        #endregion Properties

        #region Variables
        /// <summary>
        /// Level's name
        /// </summary>
        string mName = "";

        /// <summary>
        /// Vector of all WorldObjects drawn on the level.
        /// </summary>
        protected List<WorldObject> mSprites = new List<WorldObject>();

        /// <summary>
        /// This level's actors. 
        /// </summary>
        protected Dictionary<string, Actor> mActors = new Dictionary<string, Actor>();

        /// <summary>
        /// This level's Animations
        /// </summary>
        protected List<Animation> mAnims = new List<Animation>();

        #endregion Variables

        #region Methods

        #region Actions
        internal void Load()
        {
            LoadBehavior();
        }
        internal void Update()
        {
            UpdateBehavior();
        }
        #endregion Actions

        #region Draw
        internal void DrawScene(Microsoft.Xna.Framework.GameTime gameTime)
        {
            This.Game.spriteBatch.Begin();
            /** Draw BG */
            if (Background != null)
                Background.Draw();

            /** Draw Sprites*/
            DrawSprites(gameTime);

            /** Draw Boundary Data */
            DrawCollisions();
            This.Game.spriteBatch.End();
        }
        internal void DrawSprites(Microsoft.Xna.Framework.GameTime gameTime)
        {
            foreach (var sprite in mSprites)
            {
                sprite.Draw(gameTime);
            }
        }

        void DrawCollisions()
        {
            if(Game.ShowCollisions){
                Background.DrawCollisions();
                foreach (var sprite in mSprites)
                {
                    sprite.DrawCollisions();
                }
            }
        }
        #endregion Drawing

        public void AddSprite(Sprite sp)
        {
            mSprites.Add(sp);
        }

        public Sprite GetSprite(string name)
        {
            return (mSprites.Find(delegate(WorldObject s) { return s.Name == name; }) as Sprite);
        }

        public void RemoveSprite(Sprite sp)
        {
            mSprites.Remove(sp);
        }

        public void AddAnimation(Animation anim)
        {
            mAnims.Add(anim);
        }

        public Animation GetAnimation(string name)
        {
            return mAnims.Find(delegate(Animation a) { return a.Name == name; });
        }

        public void RemoveAnimation(Animation anim)
        {
            mAnims.Remove(anim);
        }

        public void AddActor(string name, Actor actor)
        {
            mActors[name]= actor;
        }

        public void RemoveActor(string name)
        {
            mActors.Remove(name);
        }
        #endregion Methods
    }
}